metricplugin/api.go

```diff
@@ -0,0 +1,81 @@
+package metricplugin
+
+import (
+	"time"
+
+	"github.com/pkg/errors"
+
+	bsmsg "github.com/ipfs/go-bitswap/message"
+	peer "github.com/libp2p/go-libp2p-core/peer"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// A BitswapMessage is the type pushed to remote clients for recorded incoming
+// Bitswap messages.
+type BitswapMessage struct {
+	// The entries requested.
+	WantlistEntries []bsmsg.Entry `json:"wantlist_entries"`
+
+	// Whether this was a full wantlist.
+	FullWantList bool `json:"full_wantlist"`
+}
+
+// ConnectionEventType specifies the type of connection event.
+type ConnectionEventType int
+
+const (
+	// Connected specifies that a connection was opened.
+	Connected ConnectionEventType = 0
+	// Disconnected specifies that a connection was closed.
+	Disconnected ConnectionEventType = 1
+)
+
+// A ConnectionEvent is the type pushed to remote clients for recorded
+// connection events.
+type ConnectionEvent struct {
+	// The multiaddress of the remote peer.
+	Remote ma.Multiaddr `json:"remote"`
+
+	// The type of this event.
+	ConnectionEventType ConnectionEventType `json:"connection_event_type"`
+}
+
+// An EventSubscriber can handle events generated by the plugin.
+type EventSubscriber interface {
+	// ID should return a unique identifier.
+	// The identifier is used to keep track of subscribers.
+	// The identifier may be reused, but only after the old use has been
+	// unsubscribed.
+	ID() string
+
+	// BitswapMessageReceived handles a Bitswap message that was recorded by the
+	// plugin.
+	// TODO for now this blocks, maybe we have to figure something out.
+	BitswapMessageReceived(timestamp time.Time, peer peer.ID, msg BitswapMessage)
+
+	// ConnectionEventRecorded handles a connection event that was recorded by
+	// the plugin.
+	// TODO for now this blocks, maybe we have to figure something out.
+	ConnectionEventRecorded(timestamp time.Time, peer peer.ID, connEvent ConnectionEvent)
+}
+
+// ErrAlreadySubscribed is returned by Subscribe if the given EventSubscriber is
+// already subscribed.
+var ErrAlreadySubscribed = errors.New("already subscribed")
+
+// PluginAPI describes the functionality provided by this plugin to remote
+// clients.
+type PluginAPI interface {
```

Comment:
Coming from the Java world, simply adding methods to the interface (and the corresponding struct) does not feel right. I don't think this is a dealbreaker right now, especially since we don't know yet which methods we want to expose. But in the future I think a command pattern or something similar would be nicer.